### ISSUE TITLE
StorageInfo.path related crash when creating class inheriting from MapTileProviderBasic

### ISSUE THREAD

**Issue by pasniak** [osmdroid/osmdroid#1613]
## Issue Type

[ ] Question
[x] Bug
[ ] Improvement

## Description and/or steps/code to reproduce the problem

I was on master for quite a while (8ac87620a9d7adbd477886983d68c4a7c05375a9) and my classes got created ok. 
I upgraded to 84c0e1fef9d4b484dca647d30ca5fd15f46e5c31 and got the below crash (in Nexus 9 API 28 emulator).
Possibly related to f27c7a473a13820d248b53f5c994687a4c228e65 ?
```
java.lang.RuntimeException: Unable to start activity ComponentInfo{com.skunkwx.pasniak.cc/com.skunkwx.pasniak.cc.ui.activity.fullscreenactivity.FullscreenActivity}: java.lang.NullPointerException: Attempt to read from field 'java.lang.String org.osmdroid.tileprovider.util.StorageUtils$StorageInfo.path' on a null object reference
	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2913)
	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3048)
	at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:78)
	at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:108)
	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:68)
	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1808)
	at android.os.Handler.dispatchMessage(Handler.java:106)
	at android.os.Looper.loop(Looper.java:193)
	at android.app.ActivityThread.main(ActivityThread.java:6669)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
Caused by: java.lang.NullPointerException: Attempt to read from field 'java.lang.String org.osmdroid.tileprovider.util.StorageUtils$StorageInfo.path' on a null object reference
	at org.osmdroid.config.DefaultConfigurationProvider.getOsmdroidBasePath(DefaultConfigurationProvider.java:253)
	at org.osmdroid.config.DefaultConfigurationProvider.getOsmdroidTileCache(DefaultConfigurationProvider.java:278)
	at org.osmdroid.config.DefaultConfigurationProvider.getOsmdroidTileCache(DefaultConfigurationProvider.java:272)
	at org.osmdroid.tileprovider.modules.SqlTileWriter.getDb(SqlTileWriter.java:736)
	at org.osmdroid.tileprovider.modules.SqlTileWriter.<init>(SqlTileWriter.java:82)
	at org.osmdroid.tileprovider.MapTileProviderBasic.<init>(MapTileProviderBasic.java:89)
	at org.osmdroid.tileprovider.MapTileProviderBasic.<init>(MapTileProviderBasic.java:70)
	at org.osmdroid.tileprovider.MapTileProviderBasic.<init>(MapTileProviderBasic.java:63)
	at org.osmdroid.tileprovider.MapTileProviderBasic.<init>(MapTileProviderBasic.java:56)
	at com.skunkwx.pasniak.cc.map.tiles.MapTileProviderCustom.<init>(MapTileProviderCustom.kt:9)
..
```

On a my laptop with AS 4.0/Emulator tiles are found ok, on my hardware tablet there is no crash but no tile appear...

### If it's a bug, version(s) of android this affects:

Api 28

### Version of osmdroid the issue relates to:

6.1.7-SNAPSHOT (84c0e1fef9d4b484dca647d30ca5fd15f46e5c31)

**Comment by monsieurtanuki** [comment]
For which Android API versions does it crash? work? doesn't crash but doesn't display anything?

**Comment by pasniak** [comment]
The crash stack is on 28 (I could repro immediately anytime).  Corrrection: the real device works (it just restarted at 0,0 lat/lon, which confused me).

**Comment by pasniak** [comment]
This occurs if the app (on that particular VM) has no permission to write in `onCreate` stack. Here mapview gets created and so is the tilecache:
```
MapView(ctx, MapTileProviderCustom(ctx) /* inherits MapTileProviderBasic*/)  
```

But I have tested fresh installs and they worked well (on all VMs at hand and this is not a new one): a handler for onRequestPermissionsResult asks user for permissions to write. If permission is granted I set location with MapViewWrapper.setCacheLocation(File(<path to cache>))

Am I supposed to do this differently now?

**Comment by monsieurtanuki** [comment]
@cbalster What do you think of that?f

**Comment by pasniak** [comment]
Below is patch which made it work again (i.e no exception when app Storage is off, on 28).
```
Index: osmdroid-android/src/main/java/org/osmdroid/tileprovider/modules/MapTileFileArchiveProvider.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- osmdroid-android/src/main/java/org/osmdroid/tileprovider/modules/MapTileFileArchiveProvider.java	(revision 84c0e1fef9d4b484dca647d30ca5fd15f46e5c31)
+++ osmdroid-android/src/main/java/org/osmdroid/tileprovider/modules/MapTileFileArchiveProvider.java	(date 1593690482000)
@@ -176,16 +176,18 @@
 
           // path should be optionally configurable
           File cachePaths = Configuration.getInstance().getOsmdroidBasePath();
-          final File[] files = cachePaths.listFiles();
-          if (files != null) {
-               for (final File file : files) {
-                    final IArchiveFile archiveFile = ArchiveFileFactory.getArchiveFile(file);
-                    if (archiveFile != null) {
-                   		 archiveFile.setIgnoreTileSource(ignoreTileSource);
-                         mArchiveFiles.add(archiveFile);
-                    }
-               }
-          }
+          if (cachePaths != null) {
+			  final File[] files = cachePaths.listFiles();
+			  if (files != null) {
+				  for (final File file : files) {
+					  final IArchiveFile archiveFile = ArchiveFileFactory.getArchiveFile(file);
+					  if (archiveFile != null) {
+						  archiveFile.setIgnoreTileSource(ignoreTileSource);
+						  mArchiveFiles.add(archiveFile);
+					  }
+				  }
+			  }
+		  }
 	}
 
 	private synchronized InputStream getInputStream(final long pMapTileIndex,
Index: osmdroid-android/src/main/java/org/osmdroid/config/DefaultConfigurationProvider.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- osmdroid-android/src/main/java/org/osmdroid/config/DefaultConfigurationProvider.java	(revision 84c0e1fef9d4b484dca647d30ca5fd15f46e5c31)
+++ osmdroid-android/src/main/java/org/osmdroid/config/DefaultConfigurationProvider.java	(date 1593689370000)
@@ -249,11 +249,14 @@
 
     @Override
     public File getOsmdroidBasePath(Context context) {
-        if (osmdroidBasePath==null)
-            osmdroidBasePath = new File(StorageUtils.getBestWritableStorage(context).path, "osmdroid");
         try {
-            osmdroidBasePath.mkdirs();
-        }catch (Exception ex){
+            if (osmdroidBasePath==null) {
+                StorageUtils.StorageInfo storageInfo = StorageUtils.getBestWritableStorage(context);
+                String pathToStorage = storageInfo.path;
+                osmdroidBasePath = new File(pathToStorage, "osmdroid");
+                osmdroidBasePath.mkdirs();
+            }
+        } catch (Exception ex){
             Log.d(IMapView.LOGTAG, "Unable to create base path at " + osmdroidBasePath, ex);
             //IO/permissions issue
             //trap for android studio layout editor and some for certain devices

```

**Comment by monsieurtanuki** [comment]
Don't hesitate to create a PR.

**Comment by cbalster** [comment]
Yeah, this is related and probably the case I described at the end in the original PR: https://github.com/osmdroid/osmdroid/pull/1571 
The tl:dr is: You have to make sure the first use of `getOsmdroidBasePath()` / `getOsmdroidTileCache()` in your codepath is one with the context supplied, so: `getOsmdroidBasePath(Context)` / `getOsmdroidTileCache(Context)`. Otherwise on API29 and above it cannot be guaranteed that any path is found. If you just instantiate a MapView this done for you by the hacky workaround, but if you use custom TileProviders, things get a little more tricky. The safe way is to just call the method with context before instantiating any mapview or TileProviders.
I took a quick look at you PR and without diving into the code, I'd say this covers your use case and prevents the crash and still works since you're setting the path manually afterwards. I'm not 100% sure this doesn't have any side effects in other use cases, but unfortunately I don't have the time right now to take a closer look.
As mentioned in the original PR this can only be solved by re-designing the storage detection mechanism to only work with a context - ideally at the same time moving away from the static singleton approach for the Configuration. But this will be a breaking change.

**Comment by monsieurtanuki** [comment]
I've just merged the PR.
@cbalster @pasniak Obviously the idea was to fix the current issue (and to do it quickly), but if you guys have suggestions for a cleaner next version please create a new issue, where we have time to discuss the different solutions - and their possible drawbacks.
